### ISSUE TITLE
Design of a general API for Regression problems

### ISSUE THREAD

**Issue by BigCrunsh** [JuliaStats/RegERMs.jl#3]
This issue carries on the discussion at https://github.com/JuliaStats/Roadmap.jl/issues/14 with the intend to make things more concrete. To start with this, I see a list of apparent questions and TODOs. 

As always your suggestions and opinions are really appreciated:

**Interfaces**
Basic functionality of a regression model should be 1) optimise the model parameters using a set of instances and labels and 2) make a prediction for a set of instances. Perhaps, we should do some renaming, e.g.:
- Model -> RegressionModel; make it abstract
- classify -> predict?
- more comprehensive name for RegERMs?

See also https://github.com/JuliaStats/Roadmap.jl/wiki/Proposal:-interface-for-JuliaStats-packages-(Statistics-and-Machine-Learning).

**Instances of RegERMs**
What should happen with certain instances like SVM, LogReg, RidgeRegression, Lasso,... ?
- implemented / instantiated in this package 
- moved to subfolder 
- own packages which are based on this package for each of these instances.

**loss.jl / regularizer.jl**
The aim of this package is to combine these concepts into the regularised empirical risk minimisation framework. For now, it makes sense to have them in this package and conflate them with existing approaches like https://github.com/johnmyleswhite/Loss.jl/.

**sgd.jl**
Stochastic gradient descent methods are typical solvers for RegERMs. I am not sure where it fits best and who else uses it. Should it be moved to (or just replaced by) https://github.com/johnmyleswhite/SGD.jl/?

**Kernels**
- Currently kernels are only implemented via mercer_map.jl. The concrete kernel functions probably should be moved into a different package, since 1) the concept of kernels is quite general can also be used by e.g., https://github.com/JuliaStats/HypothesisTests.jl and 2) there are a lot of related methods, e.g., for approximating kernels (for efficiency) which are too specific but useful to have in general.
- mercer_map.jl: computes a finite mapping w.r.t. a given kernel for a fixed training set. It feels like that this implementation of kernels belongs to the model level.

**Cross-validation**
Cross-validation is important for estimating the performance of a given model and for hyper-parameter tuning. The regularisation parameter is actually just another model parameter, which has to be optimised. Thus, procedures like marginal likelihood tuning and cross-validation should be located at the solver level. See, e.g., https://github.com/JuliaStats/MLBase.jl.

I would like to make the whole discussion a bit more specific and apply / adapt these concepts asap to an example like logistic regression (multi-class, kernelized) to see limits and problems.
